### DISCUSSION TITLE
Using webhook conditions.

### DISCUSSION THREAD

**Discussion by LilTrublMakr** [netbox-community/netbox#9683] [Q&A]
I am trying to get the hang of webhook conditions to leverage when a hook should be ran. So far, I have gotten the mode types to work, but would like to only be ran on certain tags since I am running switches from Ubiquiti and Cisco. Since the formats are a little different, I need to run different webhooks.

I tried running the following code to run the webhook but it does not trigger:
```
{
  "and": [
    {
      "attr": "mode.value",
      "value": "access"
    }, 
    {
      "attr": "tags.slug", 
      "value": "unifi"
    }
  ]
}
```
My  guess is because the tag is actually an array is the reason that the hook will not fire.

How do I go about making the condition to see if the hook contains a certain tag?

**Comment by jeremystretch** [comment]
To check for a single tag, you need to specify an [operation](https://docs.netbox.dev/en/stable/reference/conditions/#available-operations) to inform how the condition is evaluated. To check for an item within a list, for instance:

```
{
  "attr": "tags.slug", 
  "value": "unifi",
  "op": "contains"
}
```

**Reply by LilTrublMakr** [reply]
I added the operation and it is still not firing.

JSON of the interface:
```
{
    "id": 76,
    "url": "http://netbox.sudo/api/dcim/interfaces/76/",
    "display": "0/7 (7)",
    "device": {
        "id": 7,
        "url": "http://netbox.sudo/api/dcim/devices/7/",
        "display": "UniFi Switch 48 (1)",
        "name": "UniFi Switch 48"
    },
    "module": null,
    "name": "0/7",
    "label": "7",
    "type": {
        "value": "1000base-t",
        "label": "1000BASE-T (1GE)"
    },
    "enabled": true,
    "parent": null,
    "bridge": null,
    "lag": null,
    "mtu": null,
    "mac_address": null,
    "speed": null,
    "duplex": null,
    "wwn": null,
    "mgmt_only": false,
    "description": "",
    "mode": {
        "value": "access",
        "label": "Access"
    },
    "rf_role": null,
    "rf_channel": null,
    "rf_channel_frequency": null,
    "rf_channel_width": null,
    "tx_power": null,
    "untagged_vlan": {
        "id": 5,
        "url": "http://netbox.sudo/api/ipam/vlans/5/",
        "display": "Guest (20)",
        "vid": 20,
        "name": "Guest"
    },
    "tagged_vlans": [],
    "mark_connected": false,
    "cable": null,
    "wireless_link": null,
    "link_peer": null,
    "link_peer_type": null,
    "wireless_lans": [],
    "vrf": null,
    "connected_endpoint": null,
    "connected_endpoint_type": null,
    "connected_endpoint_reachable": null,
    "tags": [
        {
            "id": 2,
            "url": "http://netbox.sudo/api/extras/tags/2/",
            "display": "UniFi",
            "name": "UniFi",
            "slug": "unifi",
            "color": "9e9e9e"
        }
    ],
    "custom_fields": {
        "mainDeviceIP": "IP"
    },
    "created": "2022-06-19T07:15:51.913652-05:00",
    "last_updated": "2022-07-08T11:07:48.124738-05:00",
    "count_ipaddresses": 0,
    "count_fhrp_groups": 0,
    "_occupied": false
}
```

Conditional JSON:
```
{
  "and": [
    {
      "attr": "mode.value",
      "value": "access"
    },
    {
      "attr": "tags.slug",
      "value": "unifi",
      "op": "contains"
    }
  ]
}
```

Am I missing something obvious?

**Reply by LilTrublMakr** [reply]
@jeremystretch Sorry to ping you. I am starting to work on this and am still not able to get it working. If I take out the tag search (so it is just searching for access mode), it works. Doing just the tag search or adding the tag search results the webhook not firing even though it gets validated correctly on save of the webhook.

**Reply by candlerb** [reply]
Try `tags.slugs` instead of `tags.slug`:

```
>>> d = Device.objects.get(id=1)
>>> d
<Device: nuc1>
>>> d.tags.slug
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: '_TaggableManager' object has no attribute 'slug'
>>> d.tags.slugs
<bound method _TaggableManager.slugs of <taggit.managers._TaggableManager object at 0x7fdfbeaebd00>>
>>> d.tags.slugs()
<RestrictedQuerySet ['prom_node']>
>>> list(d.tags.slugs())
['prom_node']
>>>
```

If this fixes the problem, then I think that Netbox should have some way to report invalid attr expressions in webhook conditionals. Have you checked `journalctl -eu netbox` to see if anything is logged? If not, then perhaps it should.

**Reply by candlerb** [reply]
AFAICS, invalid names in the dotted attr path are simply ignored (`netbox/extras/conditions.py`)

```
    def eval(self, data):
        """
        Evaluate the provided data to determine whether it matches the condition.
        """
        try:
            value = functools.reduce(dict.get, self.attr.split('.'), data)
        except TypeError:
            # Invalid key path
            value = None
```

**Reply by LilTrublMakr** [reply]
Unfortunately, no difference between `tags.slug` and `tags.slugs`. I also tried `tag.slug(s)` and same thing. 

Here is the logging output:
```
[pid: 174|app: 0|req: 90/151] 192.168.0.169 () {52 vars in 1286 bytes} [Sun Sep 25 08:04:17 2022] POST /dcim/interfaces/364/edit/?return_url=/dcim/devices/46/interfaces/ => generated 0 bytes in 142 msecs (HTTP/1.1 302) 9 headers in 520 bytes (1 switches on core 0)
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/rq/worker.py", line 1068, in perform_job
    rv = job.perform()
  File "/usr/lib/python3.9/site-packages/rq/job.py", line 847, in perform
    self._result = self._execute()
  File "/usr/lib/python3.9/site-packages/rq/job.py", line 870, in _execute
    result = self.func(*self.args, **self.kwargs)
  File "/app/netbox/netbox/extras/webhooks_worker.py", line 36, in process_webhook
    if not eval_conditions(webhook, data):
  File "/app/netbox/netbox/extras/webhooks_worker.py", line 24, in eval_conditions
    if ConditionSet(webhook.conditions).eval(data):
  File "/app/netbox/netbox/extras/conditions.py", line 148, in eval
    return func(d.eval(data) for d in self.conditions)
  File "/app/netbox/netbox/extras/conditions.py", line 148, in <genexpr>
    return func(d.eval(data) for d in self.conditions)
  File "/app/netbox/netbox/extras/conditions.py", line 72, in eval
    result = self.eval_func(value)
  File "/app/netbox/netbox/extras/conditions.py", line 106, in eval_contains
    return self.value in value
TypeError: argument of type 'NoneType' is not iterable
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/rq/worker.py", line 1068, in perform_job
    rv = job.perform()
  File "/usr/lib/python3.9/site-packages/rq/job.py", line 847, in perform
    self._result = self._execute()
  File "/usr/lib/python3.9/site-packages/rq/job.py", line 870, in _execute
    result = self.func(*self.args, **self.kwargs)
  File "/app/netbox/netbox/extras/webhooks_worker.py", line 36, in process_webhook
    if not eval_conditions(webhook, data):
  File "/app/netbox/netbox/extras/webhooks_worker.py", line 24, in eval_conditions
    if ConditionSet(webhook.conditions).eval(data):
  File "/app/netbox/netbox/extras/conditions.py", line 148, in eval
    return func(d.eval(data) for d in self.conditions)
  File "/app/netbox/netbox/extras/conditions.py", line 148, in <genexpr>
    return func(d.eval(data) for d in self.conditions)
  File "/app/netbox/netbox/extras/conditions.py", line 72, in eval
    result = self.eval_func(value)
  File "/app/netbox/netbox/extras/conditions.py", line 106, in eval_contains
    return self.value in value
TypeError: argument of type 'NoneType' is not iterable
[pid: 174|app: 0|req: 91/152] 192.168.0.169 () {46 vars in 1270 bytes} [Sun Sep 25 08:04:17 2022] GET /dcim/devices/46/interfaces/ => generated 310423 bytes in 973 msecs (HTTP/1.1 200) 9 headers in 504 bytes (3 switches on core 0)
```

**Reply by candlerb** [reply]
Thanks. I guess that came from the netbox-rq worker? You can also get it via Admin > (System) Background Tasks, click on the number of Failed Jobs in the default queue, then the job ID next to "extras.webhooks_worker.process_webhook"

When I do this, it also shows the "data" item. Here's an example, reformatted slightly, of an IPAddress object:

```
data: { ...
    'status': OrderedDict([('value', 'active'), ('label', 'Active')]), ...
    'tags': [OrderedDict([('id', 9), ('url', '/api/extras/tags/9/'), ('display', 'gateway'), ('name', 'gateway'), ('slug', 'gateway'), ('color', '009688')])], ...
```

Hence "tags" is a list of OrderedDicts.

Try this patch as a proof-of-concept. You'll need to restart `netbox-rq` after applying it, and use "tags.slug" as the attribute in your condition.

```
--- a/netbox/extras/conditions.py
+++ b/netbox/extras/conditions.py
@@ -64,8 +64,13 @@ class Condition:
         """
         Evaluate the provided data to determine whether it matches the condition.
         """
+        def get(obj, key):
+            if isinstance(obj, list):
+                return [dict.get(item, key) for item in obj]
+            else:
+                return dict.get(obj, key)
         try:
-            value = functools.reduce(dict.get, self.attr.split('.'), data)
+            value = functools.reduce(get, self.attr.split('.'), data)
         except TypeError:
             # Invalid key path
             value = None
```

I've tested it and it works for me.  If it solves your problem, then it can be raised as a bug report or feature request (although it should really have a test case written as well)

**Reply by LilTrublMakr** [reply]
Yep, that fixed it for me as well. Not really sure the proper procedures to create a test case and bug report. Is that something you would be willing to take care of?

**Reply by kkthxbye-code** [reply]
@LilTrublMakr - There's already a feature request open for it: https://github.com/netbox-community/netbox/issues/9876

Maybe @candlerb is interested in suggesting his fix there.

**Reply by LilTrublMakr** [reply]
It seems the issue is about to be closed for inactivity. I honestly think this is something that should be implemented. I have been implementing the proposed fix every time that Netbox gets updated.

**Reply by AndMrzv** [reply]
Can you make it work by manually checking "tags[0].slug" OR "tags[1].slug" etc. in conditions or netbox does not support that? Having this problem with cable connections after 3.3.

**Reply by LilTrublMakr** [reply]
I believe that the problem is that the webhook conditions doesn't see embedded arrays. At the very least, it does not allow for tags and I have not tested any others. Since tags are the only thing that is an extra actual array, that makes sense. @jeremystretch seems to think it should work, but has been tested by multiple people have tested it and came to the conclusion that it does not. A proposed solution has been presented both here and in issue #9876, and multiple other solutions that I feel should be implemented but would produce breaking changes, so I can understand why they would they wouldn't.

EDIT: clarified webhook conditions

**Comment by tyler-8** [comment]
Thankful that I found this thread as I was having a heck of time getting my tag-based conditions to work and thought I was doing it wrong. I'll keep my eye on #9876
